### PR TITLE
match postgres interval handling with head/tail time units

### PR DESCRIPTION
**Pull request by pH14**
Started to look in to https://github.com/MaterializeInc/materialize/issues/7918 and found that there's some interesting and subtle INTERVAL behavior in PostgreSQL that we were missing.

PostgreSQL implicitly assumes `INTERVAL 12:34` to be H:M, which is what Materialize currently understands:

```
=# SELECT INTERVAL '12:34';
12:34:00
```

But, for reasons I could not find in the manual, PG transitions the format to M:S using the `TO` syntax:

```
=# SELECT INTERVAL '12:34' MINUTE TO SECOND;
 00:12:34
```

One can also explicitly state that the first component is Hour:

```
=# SELECT INTERVAL '12:34' HOUR;
 12:00:00

=# SELECT INTERVAL '12:34' HOUR TO MINUTE;
 12:34:00

=# SELECT INTERVAL '12:34' HOUR TO SECOND;
 12:34:00
```

As a result, this means that intervals of form '{}:{}' cannot always be assumed to be H:M, and we need to include the head time unit, if it could affect the output, when parsing the interval. This PR plumbs the head time unit through to the parsing, and adds several unit tests.

### Motivation

Fixes https://github.com/MaterializeInc/materialize/issues/7918

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR adds a release note for any [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/user/content/release-notes.md#what-changes-require-a-release-note).

### CODE REVIEW

**Comment by umanwizard** [comment]
Can we try to find out where this syntax is documented by Postgres, if anywhere?

**Comment by pH14** [comment]
> Can we try to find out where this syntax is documented by Postgres, if anywhere?

I got kinda sucked into this, so here's what I found:

* postgres special casing `MINUTE TO SECOND` to turn 2 field intervals into M:S https://github.com/postgres/postgres/blob/27b77ecf9f4d5be211900eda54d8155ada50d696/src/backend/utils/adt/datetime.c#L2581-L2587
* test cases showing this behavior https://github.com/postgres/postgres/blob/4b160492b9b8345747dfcad468eb89e9ccc0a876/src/test/regress/expected/interval.out#L680-L684
* Cockroach doing the same to follow postgres -- https://github.com/cockroachdb/cockroach/blob/72b843ffa25b4cbb8a6ff3fe66c4a089bb57d9a6/pkg/sql/sem/tree/interval.go#L243-L251

I was also reading through the `identifier qualifer` section of the ANSI SQL spec, and can kind of understand why this behavior was decided on, though I'm no expert on parsing that document.

With all of this, the final result is that `MINUTE TO SECOND` has special handling for 2 field time intervals -- I wonder if we want to change this PR to just hardcode `is_minute_to_second` through?

**Comment by pH14** [comment]
One test case was failing, and while it's a little scary editing one of the existing tests, I've verified it did not match Postgres behavior and updated the args.

With the previous version under Postgres we got:

```
psql =# SELECT extract(minute FROM INTERVAL '123:10' MINUTE TO SECOND)
ERROR:  interval field value out of range: "123:10"
```

This now returns something similar:

```
materialize=> SELECT extract(minute FROM INTERVAL '123:10' MINUTE TO SECOND);
ERROR:  invalid input syntax for type interval: MINUTE must be [-59, 59], got 123: "123:10"
```

**Comment by umanwizard** [comment]
Spoke to Paul and Nikhil offline. I think I’m convinced that the behavior is a net win because it better matches Postgres. (I haven’t looked at the code, which is more in Nikhil or Sean’s wheelhouse; just commenting on the behavior).

If you want to go the extra mile here, we should (1) make sure we have a test for all twelve cases we enumerated in our conversation today; (2) precisely document the behavior in the public-facing docs.

File: test/sqllogictest/interval.slt [hunk 1]
```diff
@@ -127,6 +127,76 @@ SELECT INTERVAL '1-2 3 4:5:6.7' DAY TO HOUR;
...
+
+query T
+SELECT INTERVAL '12:34' HOUR TO SECOND;
+----
+12:34:00
+
+query T
+SELECT INTERVAL '12:34' MINUTE TO SECOND;
+----
+00:12:34
```
Comment (sploiselle) [inline test/sqllogictest/interval.slt:163]: Would you mind throwing these tests in here, as well?

```
materialize=> SELECT INTERVAL '12:34' MONTH TO SECOND;
 interval 
----------
 12:34:00
(1 row)

materialize=> SELECT INTERVAL '12:34' DAY TO SECOND;
 interval 
----------
 12:34:00
(1 row)
```

my thinking here is ensuring that we don't somehow misconstrue the head here as specifying that the leading element is anything greater than an hour

**Review by sploiselle — APPROVED** [review]
LGTM modulo adding "foolish" tests

**Comment by pH14** [comment]
👍 thanks for taking a look, I'll add those and close this one out next Monday

**Pull request merged**